File: 0875-koko-eating-bananas/0875-koko-eating-bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
const minEatingSpeed = (piles, h) =>{
    let low = 1;
    let high = Math.max(...piles);
    let ans = high
    while (low <= high) {
        const mid = low + Math.floor((high - low)/2);
        let hours = 0;
		//Calculate total time required to finish all bananas with mid as average speed
        for(let p of piles) {
            hours += Math.ceil(p/mid)
        }
        if(hours <= h) {
		//If this is the possible ans than lesser value can also be ans so shift end to mid-1
            ans = Math.min(mid, ans);
            high = mid - 1;
        }
        else  {
            low = mid + 1;     
        }           
    }
    return ans;
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    private long computeTime(int[] piles, int k) {
        long res = 0;
        for(int pile : piles) 
            res += ((pile + k - 1) / k);
        return res;
    }

    private int bs(int[] piles, int left, int right, int h) {
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (computeTime(piles, mid) <= h) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    public int minEatingSpeed(int[] piles, int h) {
        int maxim = 0;
        for(int pile : piles) maxim = Math.max(pile, maxim);

        return bs(piles, 1, maxim, h);
    }
}